### PR TITLE
Webpack 5

### PR DESCRIPTION
**Pull request by gtrufitt**
##  What does this change?

This upgrade DCR to Webpack 5. There are a few strange bits going on but in general things seem ok...

Thing that I had to change are inline, it's likely that I'll bump some other modules and shake out any other issues before merging but it builds, which is a plus...

### CODE REVIEW

File: makefile [hunk 1]
```diff
@@ -32,7 +32,7 @@ deploy:
 
 build: clean-dist install
 	$(call log, "building production bundles")
-	@NODE_ENV=production webpack --config scripts/webpack/frontend
+	@NODE_ENV=production webpack --config ./scripts/webpack/frontend.js
```
Comment (gtrufitt) [inline makefile:35]: Webpack 5 couldn't find this script anymore...

File: package.json [hunk 2]
```diff
@@ -33,11 +33,11 @@
     },
     "dependencies": {
         "@braze/web-sdk-core": "^3.0.0",
-        "@emotion/core": "^10.0.35",
         "@cypress/skip-test": "^2.5.0",
+        "@emotion/core": "^10.0.35",
         "@guardian/ab-core": "^2.0.0",
         "@guardian/ab-react": "^2.0.1",
-        "@guardian/atoms-rendering": "^2.0.9",
+        "@guardian/atoms-rendering": "^3.0.0",
```
Comment (gtrufitt) [inline package.json:40]: Bump to get get passed Webpack not being able to bundle `react-dom/server` which kept erroring with this:

```
 ERROR in ./node_modules/@guardian/atoms-rendering/dist/index.modern.js 10:0-55
  Module not found: Error: Package path ./compat/server is not exported from package /Users/gareth_trufitt/code/dotcom-rendering/node_modules/preact (see exports field in /Users/gareth_trufitt/code/dotcom-rendering/node_modules/preact/package.json)
   @ ./src/web/components/App.tsx 113:9-115:30
   @ ./src/web/components/HydrateApp.tsx 9:0-28 41:38-41
   @ ./src/web/browser/react/init.ts 4:0-57 26:2-12
```

File: package.json [hunk 3]
```diff
@@ -64,7 +64,6 @@
         "ajv": "^6.10.1",
         "aws-sdk": "^2.791.0",
         "chalk": "^4.1.0",
-        "columnify": "^1.5.4",
```
Comment (gtrufitt) [inline package.json:67]: I don't remember deleting this or filesizegzip but they're not used so cool

File: scripts/frontend/dev-server.js [hunk 4]
```diff
@@ -39,9 +39,7 @@ const go = () => {
     app.use(
         webpackDevMiddleware(compiler, {
             serverSideRender: true,
```
Comment (gtrufitt) [inline scripts/frontend/dev-server.js:41]: Bot of these are deprecated in `webpackDevMiddleware`

File: scripts/webpack/browser.js [hunk 5]
```diff
@@ -55,8 +55,6 @@ module.exports = ({ isLegacyJS }) => ({
     },
     // fix for known issue with webpack dynamic imports
     optimization: {
```
Comment (gtrufitt) [inline scripts/webpack/browser.js:57]: Deprecated

File: scripts/webpack/browser.js [hunk 6]
```diff
@@ -76,15 +73,17 @@ module.exports = ({ isLegacyJS }) => ({
...
+                exclude: {
+                    and: [/node_modules/],
+                    not: [
+                        // Include all @guardian modules, except automat-modules
+                        /@guardian\/(?!(automat-modules))/,
+
+                        // Include the dynamic-import-polyfill
+                        /dynamic-import-polyfill/,
+                    ],
+                },
```
Comment (gtrufitt) [inline scripts/webpack/browser.js:85]: Please look closely at this, I think it is equivalent in the new syntax but it's a bit weird still.

File: scripts/webpack/frontend.js [hunk 7]
```diff
@@ -17,8 +17,8 @@ const commonConfigs = ({ platform }) => ({
     stats: 'errors-only',
     devtool:
         process.env.NODE_ENV === 'production'
-            ? 'sourcemap'
-            : 'cheap-module-eval-source-map',
+            ? 'source-map'
```
Comment (gtrufitt) [inline scripts/webpack/frontend.js:20]: `sourcemap` was never a thing 🤷‍♂️

File: scripts/webpack/server.js [hunk 8]
```diff
@@ -36,7 +35,10 @@ module.exports = () => ({
         rules: [
             {
                 test: /(\.tsx|\.js|\.ts)$/,
-                exclude: /node_modules\/(?!(@guardian\/discussion-rendering)|(@guardian\/types)|(dynamic-import-polyfill))\/.*/,
+                exclude: {
+                    and: [/node_modules/],
+                    not: [/@guardian/, /dynamic-import-polyfill/],
+                },
```
Comment (gtrufitt) [inline scripts/webpack/server.js:41]: Please check equivalence (though I did change it - we are looking at all `@guardian` modules on the server

**Review by gtrufitt — COMMENTED** [review]
https://webpack.js.org/blog/2020-10-10-webpack-5-release/

**Comment by liywjl** [comment]
It looks good but you are stil getting some unusual CI issues

**Comment by JamieB-gu** [comment]
Equivalent PR on AR: https://github.com/guardian/apps-rendering/pull/927

FYI @dskamiotis

**Comment by gtrufitt** [comment]
The braze changes above 

| Filename | Size | Change | |
|:--- |:---:|:---:|:---:|
| `dist/vendors~braze-web-sdk-core.js` | 0 B | -34.9 kB (0%) |  |
| `dist/vendors~braze-web-sdk-core.legacy.js` | 0 B | -34.9 kB (0%) |  |
| `dist/vendors~guardian-braze-components.js` | 0 B | -17.3 kB (0%) |  |
| `dist/vendors~guardian-braze-components.legacy.js` | 0 B | -17.4 kB (0%) |  |

became 

| Filename | Size | Change | |
|:--- |:---:|:---:|:---:|
| `dist/braze-web-sdk-core.js` | 34.8 kB | 0 B |  |
| `dist/braze-web-sdk-core.legacy.js` | 34.8 kB | 0 B |  |
| `dist/guardian-braze-components.js` | 17.1 kB | 0 B |  |
| `dist/guardian-braze-components.legacy.js` | 17.1 kB | 0 B |  |

**Review by liywjl — APPROVED** [review]
LGTM 👍

File: package.json (hunk 2 above)
Comment (oliverlloyd) [reply package.json:40]: I guess this PR negates the problems in @liywjl 's [one here](https://github.com/guardian/dotcom-rendering/pull/2142/files) then, which is nice

File: package.json [hunk 9]
```diff
@@ -105,16 +103,15 @@
         "ophan-tracker-js": "^1.3.23",
         "pm2": "^4.5.0",
         "polished": "^1.9.2",
-        "preact": "^10.3.1",
-        "preact-render-to-string": "^5.1.10",
+        "preact": "^10.5.7",
```
Comment (oliverlloyd) [inline package.json:106]: Makes [this Snyk nag](https://github.com/guardian/dotcom-rendering/pull/2113) moot

File: scripts/webpack/browser.js [hunk 10]
```diff
@@ -76,15 +73,17 @@ module.exports = ({ isLegacyJS }) => ({
...
-                        exclude: [
-                            /@guardian\/(?!(automat-modules))/,
-                            /dynamic-import-polyfill/,
-                        ],
-                    },
-                ],
+                exclude: {
+                    and: [/node_modules/],
+                    not: [
+                        // Include all @guardian modules, except automat-modules
```
Comment (oliverlloyd) [inline scripts/webpack/browser.js:79]: Unless I'm confused (v possible) shouldn't this read // Exclude all..., not Include

File: scripts/webpack/browser.js [hunk 11]
```diff
@@ -76,15 +73,17 @@ module.exports = ({ isLegacyJS }) => ({
...
-                    {
-                        test: /node_modules/,
-                        exclude: [
-                            /@guardian\/(?!(automat-modules))/,
-                            /dynamic-import-polyfill/,
-                        ],
-                    },
-                ],
+                exclude: {
+                    and: [/node_modules/],
```
Comment (oliverlloyd) [inline scripts/webpack/browser.js:77]: Do you need this `and`? I mean, node_modules and what?

File: scripts/webpack/frontend.js (hunk 7 above)
Comment (oliverlloyd) [reply scripts/webpack/frontend.js:20]: Maybe we don't need source maps?

File: scripts/webpack/server.js [hunk 12]
```diff
@@ -13,12 +13,11 @@ module.exports = () => ({
     target: 'node',
     optimization: {
         minimize: false,
-        namedModules: true,
         runtimeChunk: false,
     },
     externals: [
         require('webpack-node-externals')({
-            whitelist: [/^@guardian/],
+            allowlist: [/^@guardian/],
```
Comment (oliverlloyd) [inline scripts/webpack/server.js:20]: 👏 Webpack

File: scripts/webpack/server.js (hunk 8 above)
Comment (oliverlloyd) [reply scripts/webpack/server.js:41]: Webpack says:

> Try to avoid exclude and prefer include

Can we reverse this ^^?

Comment (oliverlloyd) [reply scripts/webpack/server.js:41]: That said, i exclude makes sense for node_modules

File: src/web/components/App.tsx [hunk 13]
```diff
@@ -65,6 +65,7 @@ import { trackPerformance } from '../browser/ga/ga';
 // *******************************
 // ****** Dynamic imports ********
 // *******************************
+
```
Comment (oliverlloyd) [inline src/web/components/App.tsx:68]: 🤷 but okay

**Review by oliverlloyd — APPROVED** [review]
Awesome and very exciting work!

File: scripts/webpack/browser.js (hunk 10 above)
Comment (gtrufitt) [reply scripts/webpack/browser.js:79]: `exclude not` means `include`.  Tbh, this was harder to understand and test than the regex, so if I can do it a simpler way, for sure, I'd like to!

**Comment by oliverlloyd** [comment]
Just a late night thought, will this change affect the `webpack.config.js` file in the `/.storybook` directory? If we can get [SB v6 working](https://storybook.js.org/blog/storybook-6-migration-guide/) we can [remove the typescript config](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md#zero-config-typescript) from here but a lot of it will need to remain.

**Comment by dskamiotis** [comment]
> Just a late night thought, will this change affect the `webpack.config.js` file in the `/.storybook` directory? If we can get [SB v6 working](https://storybook.js.org/blog/storybook-6-migration-guide/) we can [remove the typescript config](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md#zero-config-typescript) from here but a lot of it will need to remain.

We are currently experiencing some problems with upgrading to Webpack 5 due to Storybook in Apps-Rendering - we are currently using Storyboard v6. (https://github.com/guardian/apps-rendering/pull/927)

File: package.json (hunk 2 above)
Comment (liywjl) [reply package.json:40]: I used this PR as a base for the Youtube changes. Looks like yes this PR has fixed the `jest` issues I had 🎊 
Why or how? 🤷  Thats beyond me. 

Thanks a million Garth!!!

**Pull request merged**